### DISCUSSION TITLE
vcpkg seems to randomly rebuild all packages when running the CI?

### DISCUSSION THREAD

**Discussion by Pentarctagon** [microsoft/vcpkg#25525] [Q&A]
For the Battle for Wesnoth we're currently using the vcpkg+cmake integration to manage Wesnoth's dependencies on Windows.

These dependencies take quite a while to build - almost an hour, and about twice as long as building Wesnoth itself - so for our Windows CI jobs (with Github Actions) I enabled caching for the built vcpkg dependencies. The problem is that vcpkg seems to decide more or less at random that it needs to rebuild all the dependencies, which then defeats the purpose of caching them. For example (the Windows jobs are the bottom two):
* https://github.com/wesnoth/wesnoth/actions/runs/2580439823 - both the Release and Debug jobs correctly used the cached dependencies, since the `Use cmake` step took under two minutes.
* https://github.com/wesnoth/wesnoth/actions/runs/2580718914 - the CI run for the next commit rebuilds all dependencies for the Debug job, but not the Release job, since the `Use cmake` step took ~52 minutes.
* https://github.com/wesnoth/wesnoth/actions/runs/2582800845 - the CI run for the commit after that is then back to correctly using the cached dependencies.

Important things to note:
* The cache is **not updated** after any of these CI runs. All three used the exact same cache.
* vcpkg itself is always checked out to the [same commit](https://github.com/wesnoth/wesnoth/blob/348834b03ae63833628fc5354a4682ce41f4462c/.github/workflows/ci-main.yml#L288-L291) each time.

The question then: Why is cmake/vcpkg only sometimes deciding it needs to rebuild all the dependencies, and how can I get it to stop doing that?

**Comment by Osyotr** [comment]
You can check difference between `vcpkg_abi_info.txt` files in some package archive to see exactly what caused cache invalidation.
The most common cause for me was compiler tracking and external tools update.

**Reply by Pentarctagon** [reply]
Wouldn't that cause all subsequent CI runs to rebuild all the dependencies for both Windows jobs though? Rather than

Debug: use cache -> cache invalid -> use cache
Release: use cache -> use cache -> use cache
